File: src/main/java/by/auction/service/CategoryService.java
package by.auction.service;

import by.auction.entity.Category;

import java.util.List;
import java.util.Optional;

/**
 * Service for data management in a table "categories"
 */
public interface CategoryService {

    /**
     * Find all categories
     * @return list of found categories
     */
    List<Category> findAll();

    /**
     * Find category by name
     * @param name
     * @return Optional of found category
     */
    Optional<Category> findByName(String name);

    /**
     * Save category in DB
     * @param category
     * @return saved category
     */
    Category save(final Category category);

    /**
     * Delete category by ID
     * @param name
     */
    void deleteByName(String name);

}

File: auction-client-angular/src/app/services/category/category.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Observable} from 'rxjs/Observable';
import {Category} from '../../models/category';
import {ConfigService} from "../config/config.service";

/**
 * Service to access categories data
 */
@Injectable()
export class CategoryService {

  /**
   * Constructor for category service
   * @param {HttpClient} http
   * @param {ConfigService} config
   */
  constructor(private http: HttpClient,
              private config: ConfigService) {
  }

  /**
   * Get api uri
   * @returns {string}
   */
  uri() {
    return this.config.getApiHref() + 'categories';
  }

  /**
   * Find all categories
   * @returns {Observable<Category[]>}
   */
  findAll(): Observable<Category[]> {
    return this.http.get<Category[]>(this.uri());
  }

  /**
   * Save category
   * @param {Category} category
   * @returns {Observable<Category>}
   */
  save(category: Category): Observable<Category> {
    return this.http.post<Category>(this.uri(), category);
  }

  /**
   * Delete category
   * @param {Category} category
   * @returns {Observable<Object>}
   */
  deleteCategory(category: Category) {
    return this.http.delete(this.uri(), {
      params: new HttpParams().set('delete', category.name)
    });
  }

}
